client/main.go

```diff
@@ -49,11 +50,22 @@ func main() {
 	if len(*dataDirPtr) == 0 {
 		log.Fatal("-datadir argument required")
 	}
-	if len(*tlsCertPtr) != 0 && len(*tlsKeyPtr) == 0 {
-		log.Fatal("-tlskey argument missing")
-	}
-	if len(*tlsCertPtr) == 0 && len(*tlsKeyPtr) != 0 {
-		log.Fatal("-tlscert argument missing")
+
+	// if accepting connections, and a TLS certificate is passed, validate
+	var tlsCertKeyPair tls.Certificate
+	if !*noAcceptPtr {
+		if len(*tlsCertPtr) != 0 {
+			if len(*tlsKeyPtr) == 0 {
+				log.Fatal("-tlskey argument missing")
+			}
+			var err error
+			tlsCertKeyPair, err = tls.LoadX509KeyPair(*tlsCertPtr, *tlsKeyPtr)
```

Comment:
I would probably change `tlsCertKeyPair` here to be `_` and not save the result. You just want to know if the paths point to a valid cert and key. Then I'd leave the rest of the code as it was. It's slightly weird adjusting the server TLS config but then still loading the cert and key from files anyway with the call to `ListenAndServeTLS`. Make sense?

Reply:
The intention of passing the `tls.Certificate` object is to avoid loading the files from disk a second time (making the assumption that disk IO is expensive). The library code (at `ServeTLS`) will examine the Certificates array of the server configuration (which in this case is populated with the already loaded `tlsCertKeyPair`) before attempting to load a certificate from the supplied strings. Only if this array is empty will it call `tls.LoadX509KeyPair` with the paths generated by  `generateSelfSignedCertAndKey` in the Peer Manager. In this way certificate files are only read from disk (or other storage) once, whether a key is provided or not. It does, however, put the responsibility of generating a Certificate object (even a blank one) on the code that calls the Peer Manager, but I'd argue that cost difference is negligible compared to having to pass two empty strings.

I admit that, in the end, the two methods are functionally equivalent. I'm open to more discussion if you still don't like the idea of changing the Peer Manager.